### PR TITLE
fixed Rultor login prefix not recognized correctly in comments, when …

### PR DESCRIPTION
**Pull request by original-brownbear**
…delimited by comma, by using regex for #821

### CODE REVIEW

**Comment by alex-palevsky** [comment]
@original-brownbear Many thanks for the PR, let me find a reviewer for it

**Comment by alex-palevsky** [comment]
@mkordas review this pull request plz

**Comment by mkordas** [comment]
@original-brownbear I'm on it

File: src/main/java/com/rultor/agents/github/qtn/QnReferredTo.java [hunk 1]
```diff
@@ -82,20 +84,28 @@ public QnReferredTo(final String self, final Question qtn) {
...
-                true,
-                String.format(
-                    QnReferredTo.PHRASES.getString("QnReferredTo.mentioned"),
-                    prefix
-                )
-            );
-            Logger.info(this, "mention found in #%d", comment.issue().number());
-            req = Req.DONE;
+        final Matcher matcher = Pattern.compile(
+            String.format(".*\\b?(%s\\b).*", prefix)
```
Comment (mkordas) [inline src/main/java/com/rultor/agents/github/qtn/QnReferredTo.java]: @original-brownbear here you have some complex regexp, but I see just one test

File: src/test/java/com/rultor/agents/github/qtn/QnReferredToTest.java [hunk 2]
```diff
@@ -76,6 +76,24 @@ public void buildsRequest() throws Exception {
     }
 
     /**
+     * QnReferredTo recognizes mention delimited by comma.
```
Comment (mkordas) [inline src/test/java/com/rultor/agents/github/qtn/QnReferredToTest.java]: @original-brownbear second word in test Javadoc should be `can`

File: src/test/java/com/rultor/agents/github/qtn/QnReferredToTest.java [hunk 3]
```diff
@@ -76,6 +76,24 @@ public void buildsRequest() throws Exception {
...
+     * @throws Exception In case of error.
+     */
+    @Test
+    public void recognizeCommaAsDelimiter() throws Exception {
```
Comment (mkordas) [inline src/test/java/com/rultor/agents/github/qtn/QnReferredToTest.java]: @original-brownbear should  be `recognizes`

File: src/test/java/com/rultor/agents/github/qtn/QnReferredToTest.java [hunk 4]
```diff
@@ -76,6 +76,24 @@ public void buildsRequest() throws Exception {
...
+        final Repo repo = new MkGithub().randomRepo();
+        final Issue issue = repo.issues().create("", "");
+        final String login = "xx";
+        issue.comments().post(String.format("hello @%s, deploy", login));
```
Comment (mkordas) [inline src/test/java/com/rultor/agents/github/qtn/QnReferredToTest.java]: @original-brownbear are you sure about that `hello `?

File: src/main/java/com/rultor/agents/github/qtn/QnReferredTo.java [hunk 5]
```diff
@@ -82,20 +84,28 @@ public QnReferredTo(final String self, final Question qtn) {
...
+        ).matcher(comment.body().trim());
+        if (matcher.matches()) {
+            if (matcher.start(1) == 0) {
+                req = this.origin.understand(comment, home);
+            } else {
+                new Answer(comment).post(
+                    true,
+                    String.format(
+                        QnReferredTo.PHRASES
+                            .getString("QnReferredTo.mentioned"),
```
Comment (mkordas) [inline src/main/java/com/rultor/agents/github/qtn/QnReferredTo.java]: @original-brownbear please put `"QnReferredTo.mentioned"` to separate line and move `.getString` line above

**Comment by mkordas** [comment]
@original-brownbear please see a few comments above

File: src/main/java/com/rultor/agents/github/qtn/QnReferredTo.java (hunk 1 above)
Comment (original-brownbear) [inline src/main/java/com/rultor/agents/github/qtn/QnReferredTo.java]: @mkordas  Added additional testcases and actually corrected the regex. \b is not a valid delimiter to the left e.g. ,@original-brownbear contrary to my assumption does not work :)

File: src/main/java/com/rultor/agents/github/qtn/QnReferredTo.java (hunk 5 above)
Comment (original-brownbear) [inline src/main/java/com/rultor/agents/github/qtn/QnReferredTo.java]: @mkordas done

File: src/test/java/com/rultor/agents/github/qtn/QnReferredToTest.java (hunk 4 above)
Comment (original-brownbear) [inline src/test/java/com/rultor/agents/github/qtn/QnReferredToTest.java]: This is the general mention case I know the deploy doesn't necessarily make a lot of added sense compared to some random word I guess right?
I left this in now as it was in the original tests but added an additional "non-hello" case to the mix, tbh not 100% sure about the readability of this though.

File: src/test/java/com/rultor/agents/github/qtn/QnReferredToTest.java (hunk 3 above)
Comment (original-brownbear) [inline src/test/java/com/rultor/agents/github/qtn/QnReferredToTest.java]: @mkordas  done :)

File: src/test/java/com/rultor/agents/github/qtn/QnReferredToTest.java (hunk 2 above)
Comment (original-brownbear) [inline src/test/java/com/rultor/agents/github/qtn/QnReferredToTest.java]: @mkordas done

File: src/main/java/com/rultor/agents/github/qtn/QnReferredTo.java (hunk 1 above)
Comment (original-brownbear) [inline src/main/java/com/rultor/agents/github/qtn/QnReferredTo.java]: @mkordas now this should be handled too.

**Comment by original-brownbear** [comment]
@mkordas I think I addressed all points now.

Made it match like this now:
* comma or whitespace in front of mention => match but don't understand as the mention is never in the beginning of the comment due to using trim
* rest as expected look for actual \b at the end of the string as the name always ends on an alphanumeric char

**Comment by mkordas** [comment]
@original-brownbear I'm on it again

File: src/main/java/com/rultor/agents/github/qtn/QnReferredTo.java [hunk 6]
```diff
@@ -82,20 +84,29 @@ public QnReferredTo(final String self, final Question qtn) {
...
-                true,
-                String.format(
-                    QnReferredTo.PHRASES.getString("QnReferredTo.mentioned"),
-                    prefix
-                )
-            );
-            Logger.info(this, "mention found in #%d", comment.issue().number());
-            req = Req.DONE;
+        final Matcher matcher = Pattern.compile(
+            String.format("(?:^|(?:.*?(?:\\s|,)))(%s)\\b.*?", prefix)
```
Comment (mkordas) [inline src/main/java/com/rultor/agents/github/qtn/QnReferredTo.java]: @original-brownbear right now this pattern is even more complex - can you extract it to constant and add a Javadoc with explanation to it?

File: src/main/java/com/rultor/agents/github/qtn/QnReferredTo.java [hunk 7]
```diff
@@ -82,20 +84,29 @@ public QnReferredTo(final String self, final Question qtn) {
...
+        ).matcher(comment.body().trim());
+        if (matcher.matches()) {
+            if (matcher.start(1) == 0) {
+                req = this.origin.understand(comment, home);
+            } else {
+                new Answer(comment).post(
+                    true,
+                    String.format(
+                        QnReferredTo.PHRASES.getString(
+                                "QnReferredTo.mentioned"
```
Comment (mkordas) [inline src/main/java/com/rultor/agents/github/qtn/QnReferredTo.java]: @original-brownbear please move back 4 spaces

Comment (original-brownbear) [inline src/main/java/com/rultor/agents/github/qtn/QnReferredTo.java]: @mkordas done

File: src/main/java/com/rultor/agents/github/qtn/QnReferredTo.java (hunk 6 above)
Comment (original-brownbear) [inline src/main/java/com/rultor/agents/github/qtn/QnReferredTo.java]: @mkordas done, hope it's not too verbose.

File: src/main/java/com/rultor/agents/github/qtn/QnReferredTo.java [hunk 8]
```diff
@@ -54,6 +56,16 @@
 public final class QnReferredTo implements Question {
 
     /**
+     * Regex that matches the @ and Rultor user login in Github comments.
+     * Matches the login when at the beginning of the comment string or when
+     * preceded by a space or comma.
+     * Login has to be bound by a word boundary to the right.
+     * Only captures the @ sign and login.
```
Comment (mkordas) [inline src/main/java/com/rultor/agents/github/qtn/QnReferredTo.java]: @original-brownbear looks great for me, thanks a lot!

File: src/main/java/com/rultor/agents/github/qtn/QnReferredTo.java [hunk 9]
```diff
@@ -54,6 +56,16 @@
 public final class QnReferredTo implements Question {
 
     /**
+     * Regex that matches the @ and Rultor user login in Github comments.
+     * Matches the login when at the beginning of the comment string or when
+     * preceded by a space or comma.
```
Comment (mkordas) [inline src/main/java/com/rultor/agents/github/qtn/QnReferredTo.java]: @original-brownbear I'm not sure whether we need to detect e.g. `,@rultor`, but it doesn't hurt either

**Comment by mkordas** [comment]
@original-brownbear good job

**Comment by mkordas** [comment]
@rultor merge

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by yegor256** [comment]
@rultor merge again

**Comment by yegor256** [comment]
@rultor merge

**Comment by mkordas** [comment]
@yegor256 please see https://github.com/yegor256/rultor/pull/1002, this should help with making merge work again

**Comment by mkordas** [comment]
@yegor256 please take care of this PR as well

**Comment by yegor256** [comment]
@rultor merge

**Comment by mkordas** [comment]
@original-brownbear can you close this PR?

**Comment by mkordas** [comment]
@rultor deploy

**Comment by original-brownbear** [comment]
@mkordas jup all done here : )

**Comment by alex-palevsky** [comment]
@mkordas since quality is good, I just added 10 mins to @original-brownbear (our architect) in transaction AP-37D02265W46925311

**Comment by mkordas** [comment]
@original-brownbear are you new architect here? :)

**Comment by original-brownbear** [comment]
@mkordas yes :)

**Comment by mkordas** [comment]
@original-brownbear congrats! I think the biggest challenge would be to improve stability of Rultor in production.

**Comment by original-brownbear** [comment]
@mkordas thanks :), 100% agreed:
Fortunately the weak points are fairly obvious :)
 * dependencies are not clearly defined version wise
 * direct/unstable interaction with the Docker daemon via CLI instead of Docker HTTP API

I already have experience building integration tests with much more complicated docker setups and made tickets for the dependencies, this should be very doable :)

**Pull request closed without merging**